Workers

igor47 opened the pull request:
this PR closes #20 

Here, I do an initial pass to implement [rq](http://python-rq.org/), a simple delayed job worker system.

RQ uses redis for communication. I've already spun up a `Cloud Memorystore` instance, which is a hosted Redis v4.0 at a cost of $30 a month, and pushed the relevant credentials using `inv.k8s-secret`. I also added a redis instance to our `docker-compose` file, so that if you now run the (misnamed) `inv run.mysql` task you'll get both MySQL and Redis in your local environment.

RQ looks at the `IN_PRODUCTION` environment variable, and if it's not `True` it will simply execute jobs right in the process. However, if the environment variable is `True`, the details of the task will be written into Redis instead of being executed. A process started with `inv rq.worker` is necessary to process through those tasks. I've created a deployment for this task (in `worker-deployment.yaml`), which will be created as part of the normal `inv image` run. To make this easier, I factored the biggest hunk of code in the existing `deployment.yaml` file -- the environment variables -- into it's own file (`container_environment.yaml`), which is nice because that's the thing we tend to modify most often anyway.

I created an object hierarchy for jobs in the `jobs` folder. We have a `BaseJob` which abstracts dealing with RQ. I also refactored `send_email` into an `EmailJob`. You can call `EmailJob.perform` to send an email immediately, or `EmailJob.enqueue` to have the email sent in a worker process.

I picked RQ because it seemed simple and there is not a lot of code, and because it claims to be inspired by [Resque](https://github.com/resque/resque) with which I have more experience than I care to admit. I also looked at [TaskTiger](https://github.com/closeio/tasktiger), [Celery](http://www.celeryproject.org/), and [Huey](https://github.com/coleifer/huey) -- I found huey last, it's by the same guy who wrote PeeWee, and I probably would've chosen it instead if I had found it sooner before I went pretty far down implementing RQ.

We have the additional requirement to implement delayed jobs -- e.g., send an email one week after launching a mission. [RQ Scheduler](https://github.com/rq/rq-scheduler) seems like a simple way to implement this functionality. I plan to add this in a follow-up PR, and add additional methods to `BaseJob` like `enqueue_at` and `enqueue_in`.

We also discussed scheduling jobs at specific times (like 2 pm on tuesdays) -- RQ scheduler has this functionality, but I'm pretty confused about how it works and I would prefer to just use cronjobs for this. I couldn't find a place where we need it right now, though, so I'm going to punt until we needed (but tell me if we need it!).

I really want to start being able to deploy potentially-broken stuff without breaking the live site -- this is getting increasingly more difficult to test locally with all the credentials and environment-specific differences. We could really use a test and a prod version of the site, and I might do that next.

igor47 committed:
requirements: add RQ

via `pip install rq`

igor47 committed:
redis credentials in app.config

igor47 committed:
add redis store instance

this instance does it's own connection pool management

igor47 committed:
jobs: initial base job using RQ

igor47 committed:
send emails in a background job

we rewrite the email function as a job and enqueue the job instead of
sending the email

igor47 committed:
tasks: add rq info and worker tasks

i also modified the `gunicorn` task to force the app into production
mode so i can force queueing of tasks

igor47 committed:
more labels

our little app is growing up and adding additional components. lets
label the existing stuff with the app, but make the webserver components
labeled with the webserver label. the service and deployment pod
selectors will now only match the webserver components

igor47 committed:
move container environment into separate file

this makes sense for frequent modification purposes (you're more often
modifying the environment than anything else about the deployment) but
also we need the environment for several different deployments

igor47 committed:
add a worker deployment

this runs an RQ worker process in the image. we make the default docker
command just list the available `inv` tasks, and we use kubernetes args
to specify what `inv` should do inside the image

igor47 committed:
log the environment we launch in

jeredw commented on Dockerfile:
```diff
@@ -32,6 +32,7 @@ COPY . $CODE_DIR
 # set this to 'production' in production deploys
 ENV ENVIRONMENT dev
```
Is this redundant with the new IN_PRODUCTION environment variable? Confusingly, I think there is also FLASK_ENV.

igor47 replied:
mmm yeah this isn't really used anywhere, might as well remove it.

jeredw commented on spaceship/config.py:
```diff
@@ -4,6 +4,8 @@
 
 class Config:
   IN_PRODUCTION = bool(os.environ.get('IN_PRODUCTION', False))
```
May be clearer to just call this USE_WORK_QUEUE. This way we can test that part out, without having to also think about what else is turned on by IN_PRODUCTION.

igor47 replied:
good call! will do.

jeredw commented on spaceship/jobs/email_job.py:
```diff
@@ -0,0 +1,23 @@
+class EmailJob(BaseJob):
+  @classmethod
+  def perform(cls, to_emails, subject, html_content, from_email='gaia@spaceshipearth.org'):
+    message = Mail(
```
Confusingly I found the Mail constructor can throw if sendgrid key is not set up properly.

igor47 replied:
huh i haven't seen that in my environment...

jeredw commented on spaceship/jobs/base_job.py:
```diff
@@ -0,0 +1,53 @@
+    )
+
+    job = q.enqueue_call(
+      func=cls.perform,
```
So to check my understanding, cls.perform will be called by an rq worker, which is a different process, right? And so this is a serialization of the call; globals etc from the environment are not available from job classes?

igor47 replied:
yeah you have to recreate the globals state, which mostly happens when you require anything from `spaceship` since the `__init__` is then loaded which does a bunch of stuff including parsing the environment and setting up DBs and logging

jeredw replied:
Hopefully our peewee models will serialize correctly, I can see us trying to pass those around.

igor47 replied:
i think it's just serializing the call and the parameters, not the code that will be invoked.

jeredw replied:
Right yeah I'm wondering what happens when the parameter is a peewee object. Do we pass... the full current state of the db object at call time? Enough info to refetch it? Does an orm cache in the worker context hold on to stale models and cause consistency issues? What if a migration happens with pending jobs that reference old models... There are just all kinds of exciting ways this can break. :)

igor47 replied:
i was wrong, it does try to serialize class instances. we should probably avoid doing that...

igor47 replied:
yeah i'm not a fan of the pickling of instances, i assumed it serialized simple data along with references by name. i'll see if i can constrain the world a bit here

jeredw commented on spaceship/jobs/email_job.py:
```diff
@@ -0,0 +1,23 @@
+    if app.config['IN_PRODUCTION']:
+      sg = SendGridAPIClient(app.config['SENDGRID_KEY'])
+      response = sg.send(message)
+    else:
```
Does this mean we will no longer send any emails in dev? Before, we would do so if the key was set up properly. This seems like an ok change but may be surprising for igork.

igor47 replied:
yup, will change back to how it was, the `IN_PRODUCTION` env var is too broad an an exception handler seems better.

jeredw commented on spaceship/views.py:
```diff
@@ -353,10 +353,13 @@ def register():
-          subject='Please verify you email for Spaceship Earth',
-          html_content=render_template('confirm_email.html',
-          confirmation_url=url_for('confirm_email', token=token, _external=True)))
+      EmailJob.enqueue(
```
(After factoring out business logic I think we may end up in a state like "emails.send_account_verification(...)", so prolly the view code won't really know anything about jobs.)

igor47 replied:
:+1: to factoring the business logic out of the views. i am still in favor of broader classes for flows, rather than individual functions. for instance, there's an account onboarding flow which sends an account verification email, and *it's the only thing which sends those emails*. so there's no good reason to expose a `send_account_verification` method in a grab-bag module called `emails`. also, there's no reason to have to open a random `emails` file when you're working on the account onboarding flow.

i would be in favor of, like, a class that's like `AccountOnboarding.perform(...)` and the `send_account_verification` function lives in that class along with other functions that have to do with account onboarding.

jeredw replied:
Hmm yeah account_onboarding may make more sense. 🙂 But then there’s still views.py, which is horizontal not vertical... and that has a certain charm because views are all peer level concerns that connect a lot. anyways happy to do what works best. 😊

also I would be remiss if I didn’t link this great yegge rant here, lol:
http://steve-yegge.blogspot.com/2006/03/execution-in-kingdom-of-nouns.html?m=1

jeredw commented on tasks/k8s.py:
```diff
@@ -35,8 +35,28 @@ def mysql_secret(ctx, host, password, port=3306, username = 'spaceship-app', db
+    'db': "Number of the Redis DB (default: '0')",
+  }
+)
+def redis_secret(ctx, host, port=3306, db=0):
```
Ohhh, _this_ is how we do this... I spent a long time googling around for what ${eval} meant when creating the secrets for oauth. I should add a method to recreate that, here.

igor47 replied:
what `{$eval: 'stuff'}` means: https://taskcluster.github.io/json-e/#Language/-eval-

jeredw commented on tasks/run.py:
```diff
@@ -43,21 +43,26 @@ def shell(ctx):
+    'force_prod': 'Set the IN_PRODUCTION variable to emulate production',
+  }
+)
+def gunicorn(ctx, force_prod=False):
```
Should I be using gunicorn instead of just inv.run? I'm confused about this now.

igor47 replied:
you're not supposed to deploy the built-in flask web server in production. in production we run `gunicorn`, and you can run it locally, too, but you'll probably be happier with `inv run` and it's interactive debugger and automatic code reloading

jeredw reviewed (approved):
Thanks, I think I get the gist of this though igork prolly has more direct experience and I think was working on a mail job queueing system, so may be good for him to look too.

igor47 commented:
thanks for the review, i'll address the comments before merging and maybe @igorkofman can weigh in too

igor47 committed:
refactor environment variables

the `IN_PRODCTION` var is not very descriptive, lets use flags for
relevant things. we also refactor the email sending to attempt it, even
locally, if possible.

igor47 commented:
closed in favor of #92
